examples/stm32h7-example/Core/Src/cli_setup.c

```diff
@@ -0,0 +1,117 @@
+/*
+ * cli_setup.c
+ *
+ *  Created on: Jul 7, 2023
+ *      Author: NeusAap
+ */
+
+
+// STD LIBS
+#include <stdio.h>
+#include <stdlib.h>
+
+// STM files
+#include "usart.h"
+#include "stm32h7xx_it.h"
+
+// Own CLI headers
+#include "cli_setup.h"
+#include "cli_binding.h"
+// Embedded CLI header (only one allowed per project)
+#define EMBEDDED_CLI_IMPL
+
+// UART buffers
+uint8_t UART_CLI_rxBuffer[UART_RX_BUFF_SIZE] = {0};
+struct {
+    char buffer[UART_TX_BUFFER_SIZE];
+    uint8_t index;
+} UART_CLI_txBuffer = { .index = 0 };
+
+// CLI buffer
+EmbeddedCli *cli;
+CLI_UINT cliBuffer[BYTES_TO_CLI_UINTS(CLI_BUFFER_SIZE)];
+
+// Bool to disable the interrupts, if CLI is not yet ready.
+bool cliIsReady = false;
+
+// Getter function, to keep only one instance of the EmbeddedCli pointer in this file.
+EmbeddedCli* getCliPointer(){
+	return cli;
+}
+
+// Write function used in 'setupCli()' to route the chars over UART.
+void writeCharToCli(EmbeddedCli *embeddedCli, char c){
+	uint8_t c_to_send = c;
+    HAL_UART_Transmit(UART_CLI_PERIPH, &c_to_send, 1, 100);
+}
+
+// Function to setup the configuration settings for the CLI, based on the definitions from this header file
+void setupCli(){
+  // UART interrupt
+  HAL_UART_Receive_IT (UART_CLI_PERIPH, UART_CLI_rxBuffer, UART_RX_BUFF_SIZE);
+
+  // Initialize the CLI configuration settings
+  EmbeddedCliConfig *config = embeddedCliDefaultConfig();
+  config->cliBuffer = cliBuffer;
+  config->cliBufferSize = CLI_BUFFER_SIZE;
+  config->rxBufferSize = CLI_RX_BUFFER_SIZE;
+  config->cmdBufferSize = CLI_CMD_BUFFER_SIZE;
+  config->historyBufferSize = CLI_HISTORY_SIZE;
+  config->maxBindingCount = CLI_MAX_BINDING_COUNT;
+
+  // Create new CLI instance
+  cli = embeddedCliNew(config);
+  // Assign character write function
+  cli->writeChar = writeCharToCli;
+
+  // CLI init failed. Is there not enough memory allocated to the CLI?
+  // Please increase the 'CLI_BUFFER_SIZE' in header file.
+  // Or decrease max binding / history size.
+  while (cli == NULL){
+	  HardFault_Handler();
+  }
+
+  // Un-comment to add a non-default reaction to unbound commands.
+  //cli->onCommand = onCliCommand;
+
+  // Add all the initial command bindings
+  initCliBinding();
+
+  // Init the CLI with blank screen
+  onClearCLI(cli, NULL, NULL);
+
+  // CLI has now been initialized, set bool to true to enable interrupts.
+  cliIsReady = true;
+}
+
+// STM32 UART callback function, to pass received characters to the embedded-cli
+void HAL_UART_RxCpltCallback(UART_HandleTypeDef *huart)
+{
+	if (huart == UART_CLI_PERIPH && cliIsReady){
+		HAL_UART_Receive_IT(UART_CLI_PERIPH, UART_CLI_rxBuffer, UART_RX_BUFF_SIZE);
+		char c = UART_CLI_rxBuffer[0];
+		embeddedCliReceiveChar(cli, c);
+	}
+}
+
+// Custom write function to route printf statements to the CLI.
+// Input is buffered until buffer is full, or \n is encountered to reduce HAL calls.
+void custom_write(int file, char *ptr, int len) {
```

Comment:
Note that using this function will mess up cli output. For correct printing you should use
```C
void embeddedCliPrint(EmbeddedCli *cli, const char *string);
```
That clears current line, prints given string and then writes cli line again.

Probably should mention it in function comment.

Reply:
This note I do not understand.
What does it mess up and why? 

My terminal correctly displays the invitation sign, the auto-complete functionality and places the cursor at the end of the command line, even when a printf statement (so a call to the `HAL_UART_Transmit()` func) gets in between.

How would you suggest I change it? And more importantly, why? :P Does it show up incorrectly on your terminal?

Reply:
I have created [a screen recording](https://github.com/funbiscuit/embedded-cli/assets/9746601/833d7640-e51f-4800-a5b2-8d500ca95257) on how it looks in my terminal. 
This might make explaining whats wrong easier :)

Reply:
What will happen when terminal in this state?
```
> get-l[]ed
```
Where `[]` denotes a cursor (staying on letter `e`) and `ed` is autocompletion?
The way I see your code, if printf fires at this moment (suppose with `test\n` input) terminal will become this:
```
> get-ltest
[]
```
You can try to add regular call to printf (say every 2 seconds) and see what happens when you start typing something.

I didn't see in your recording output from printf, is it there?

As for how to change this function - probablly you can't :) Just try to use embeddedCliPrint where you have such option instead of printf. That's why I said to just mention this limitation in comments.

Reply:
Ah yes. Quite an oversight from me :P I indeed did not have a printf periodically printing (so this problem would never occur). 
Great that you already had a function to overcome this problem! 

This is a video with my old [_write function](https://github.com/funbiscuit/embedded-cli/assets/9746601/d39d1cb4-94a4-46de-936a-d7510bebe672) with the periodic printf.

But since I really like the formatting abilities of printf, I have made an encapsulation `cli_printf()` function, that has a static char buffer, which adds the formatting arguments before calling `embeddedCliPrint()`.
I have changed out the old `_write()` function, for `cli_printf()`, and [here is a video](https://github.com/funbiscuit/embedded-cli/assets/9746601/a412c252-cef1-441c-9dff-ddf4d0a28b2b) showing this does work as expected. Could also opt to make it an dynamic function to calculate the exact length, but I personally don't prefer that, and I think this should be enough for a simple example :P

I've also changed the README.md accordingly, and also the code comments.

P.S. I did get bamboozled by the extra linebreak '\r\n' at first, but I found that you always add them in `embeddedCliPrint()`, so I just opted for adding a comment that there is no need to add linebreaks yourself.

Reply:
Yes, linebreak is added since we need to print invitation and current command on new line, so line break is necessary.
For me dynamic allocation is a no-no in microcontroller world :)